toolchain/check/handle_index.cpp

```diff
@@ -16,17 +21,93 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
   return true;
 }
 
+// Returns the argument values of the `IndexWith` interface, this corresponds to
+// the `SubscriptType` and the `ElementType`, if the class does not implement
+// the said interface this returns an empty array reference.
+static auto GetIndexWithArgs(Context& context, Parse::NodeId node_id,
+                             SemIR::TypeId self_id)
+    -> llvm::ArrayRef<SemIR::InstId> {
+  auto index_with_interface =
+      context.types().GetAs<SemIR::GenericInterfaceType>(
+          context.insts()
+              .GetAs<SemIR::StructValue>(
+                  context.LookupNameInCore(node_id, "IndexWith"))
+              .type_id);
+
+  for (const auto& impl : context.impls().array_ref()) {
+    auto impl_self_type_id = context.GetTypeIdForTypeInst(impl.self_id);
+    auto impl_constraint_type_id =
+        context.GetTypeIdForTypeInst(impl.constraint_id);
+
+    if (impl_self_type_id != self_id) {
+      continue;
+    }
+    auto interface_type =
+        context.types().TryGetAs<SemIR::InterfaceType>(impl_constraint_type_id);
+    if (!interface_type) {
+      continue;
+    }
+
+    if (index_with_interface.interface_id != interface_type->interface_id) {
+      continue;
+    }
+
+    return context.inst_blocks().GetOrEmpty(
+        context.specifics().Get(interface_type->specific_id).args_id);
+  }
+
+  return {};
+}
+
+// Creates SemIR to perform an index with base expression `operand_inst_id` and
+// `operand_type_id` for types that are not an array. This entails, checking if
+// the base expression implements the `IndexWith` interface, if so uses the `At`
+// associative method while printing a diagnostic if it isn't.
```

Comment:
```suggestion
// Performs an index with base expression `operand_inst_id` and
// `operand_type_id` for types that are not an array. This checks if
// the base expression implements the `IndexWith` interface; if so, uses the `At`
// associative method, otherwise prints a diagnostic.
```